src/cell_list.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+#ifndef CELL_LIST_H
+#define CELL_LIST_H
+
+#include <algorithm>
+#include <list>
+#include <vector>
+
+#include "line.h"
+#include "point.h"
+
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_iterator;
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range;
+
+/*
+ * A Cell List operates by subdividing an area into a regular grid and populating each cell
+ * with a list of the entities in that area.
+ * It is used to optimize lookup of nearby entities.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list
+{
+    public:
+        cell_list() = default;
+        void add( const point &p, T &value ) {
+            std::pair<point, T *> val( p, &value );
+            cells[p.y / CELL_SIZE][p.x / CELL_SIZE].push_back( val );
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
+            return cell_list_range<T, CELL_SIZE, MAXX, MAXY>( *this, p, distance );
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        std::vector<std::pair<point, T *>> cells[MAXY / CELL_SIZE][MAXX / CELL_SIZE];
```

Comment:
I think these divisions need to be `divide_round_up`, otherwise you're going to end up accessing outside the array.